### ISSUE TITLE
Why create your own solution / itemize the problem space?

### ISSUE THREAD

**Issue by NoahKunin** [deptofdefense/code.mil#34]
Major congrats on this! I'm quite familiar with this problem, having worked on it from many angles before finally getting a consensus solution together at other agencies I've worked at. It's hella tricky. I can cross-link later if you'd like once I dig up the previous issues. I also think most of that would fit in #14 if possible, so I'll connect into pre-existing issues where possible.

Overall though, I'm a little confused, on two points. 

1. For where and when (and really more accurately, from _whom_) there is no ability to copyright the information, and thus nothing to actually license _out_, why not simply make it clear the public domain takes effect, and leave it at that? Why even cover those materials in contract law?

2. When the DOD does have copyright, why roll you own instead of using a license already in circulation?

Not saying there isn't a good reason for either - there may be! If the problem was explained a little more fully, I think the detail level of feedback and potential pull requests would increase as well.

**Comment by tomberek** [comment]
It is becoming apparent that our explanation needs some work to be more clear.

1. Federal government employees do not have copyright on their work.
2. The [LICENSE-agreement.md](/LICENSE-agreement.md) is only meant to be a thin wrapper, a mechanism for us to re-use existing licenses like MIT, BSD, etc. Most people shouldn't have to even care about the LICENSE-agreement. For most practical purposes, the project is under the particular LICENSE chosen by the project manager in question.

We are considering another option as well. I'd appreciate your comments:

1. Explicitly release under public domain.
2. Have a Developer Certificate of Origin and a policy that contributions from the public (not federal employees) would only be accepted if they were under license XYZ (fill in a license here).
3. Over time, the project would become de-facto license XYZ due to public contributions. 
4. If no-one ever contributes, or if someone wanted only the original commit, it would be in the public domain, not XYZ.

**Comment by konklone** [comment]
@tomberek That second option eliminates the vast majority of my concerns, and the concerns I've seen expressed here. I can see the other complexities/annoyances that arise from having repositories be composed of contributions with a mix of copyright statuses. (Though that's often the case no matter what you do.) But it does get you out of the realm of novel licenses (or license-like creations), and out of using contract law to effect copyright-like restrictions.

**Comment by jbjonesjr** [comment]
How do these additional options work when a project is a mix of Federal Employees and non-Fed developers? I honestly have no clue, just raising the question.

**Comment by irl** [comment]
@tomberek +1 for the second option

It's simple and for the most part, the fact that it's public domain initially can be entirely ignored by developers assuming a license is chosen that has pretty much the same rights granted.

**Comment by marctjones** [comment]
TLDR;

1) I think I get what you guys are trying to do now (Sorry, I am a little slow to catch on I guess)
2) I think it is awesome that you want to prompt the use of FOSS licenses and try to make sure code remains free (as in freedom. ;) )
3) I agree with @konklone. Your second option seems more inline with what FOSS developers expect and would be pretty normal actually.
4) Having a contributor policy is awesome! More projects should have one before they find out the problem they could have been avoided if one existed.
5) Have you considered the pros and cons of using a Apache-style CLA or a copyright assignment instead of the DCO?
5.5) There is a difference between the contribution policy, the contribution agreement if any, and the license for the code.

-Marc
(This is not legal advice and represent my personal views. They are not the views of my employer and you are not my client just because you read this.)

Now, because I can be long winded, but the rest of this is still not legal advice!

##Nothing wrong with being Normal

@tomberek I think you will find that the other option you mentioned will gather a lot less attention, but that is a good thing for a lot of projects. A lot of FOSS developers care very deeply about licensing and will jump to experiment with how licensing is handled, but in the long term I think you will find that contributors also very consciously make choices on which projects they contribute to based on the license and contributor agreement involved. It is unlikely that all potential contributors will simply ignore the agreement's existence. (And if they do ignore it is only going to raise issues about if they truly consented to the agreement.) Projects are much more successful when they are known for the quality of the software rather then the novelty of the licensing scheme. It is always a shame to see perfectly good software go unused because of confusion about its licensing status or how the license operates. (See e.g., [JSLint](https://en.wikipedia.org/wiki/JSLint) )

As @konklone mentioned, your second option is also much closer to what many other existing FOSS projects already do, so will probably be viewed by most contributors as pretty normal. An unfortunate side effect of doing some novel licensing arrangement is that some developers will choose not to commit because they don't understand what a project is asking them to agree to. Heck many developers refuse to contribute to projects even when they do understand but simple prefer to only contribute to projects that use the licenses they prefer. It is not always just the license though, many developers want to understand the whole legal relationship they are agreeing to enter into.

Requiring a DCO is a increasingly popular option (It has obviously(?) worked for it's originating project, the Linux Kernel). It also has the added advantage of not being a contract which makes it easier for some private companies to deal with since there is no contract that the company has to be party to which often requires the legal department to get involved. Personally I think that might be more of advantage when dealing with corporate bureaucracies than in other ways; but a contract and a license/certification are different and different legal risks/obligations/rights will accrue depending on the situation. (Not that you shouldn't hire a lawyer to review them both. Support your local FOSS/IP counsel by buying them beer; the official food of the free software movement.) I would be interested to hear if you and @shawoods think that having or not having a contract in place with individual contributors is advantage or not from a government attorney's point of view.

##But nothing wrong trying to do something cool and unique either!

I actually think step 3 in your second option would be pretty novel! It might be a first for a government project to intentionally select a FOSS license by only accepting contributions that are compatible with the license the project wishes it could use if it were eligible for copyright protection. If its not the first time, I would love to hear about other examples.

I think this would be especially exciting, if a government project took the opportunity to adopt a copyleft license such as the GPL or AGPL! I have seen mentioned in other threads on in this repo that one of this group's concerns is the long-term openness of the project. The GPL and AGPL were explicitly design to guarantee that code would remain in the public domain even as it changes and receives contributions over time. (See para. 2 of the GPLv3's [Preamble](https://www.gnu.org/licenses/gpl.html) )

Unlike government works, the most prominent FOSS projects that are in the public domain are there because the developers intentionally put the project in the public domain. They remain in the public domain because the maintainer refuses to accept contributions unless the contribution is accompanied with a dedicated to the public domain. The most prominent project I can think of is probably the [SQLite project](https://sqlite.org/copyright.html). As others have expressed concern about the adequacy of public domain dedications in international contexts, I note that the public domain dedication required by SQLite appears to be focused primarily on addressing its status under U.S. Copyright law. But, what I believe is more recent effort, the CCO is an example of a more sophisticated approach to contributing to the public domain. I myself have been guilty of referring to the CCO simplistically as a public domain dedication, but it is actually much more complicated then that and designed to address a increasingly global world/Internet. The [legal text of the CCO](https://creativecommons.org/publicdomain/zero/1.0/legalcode) contains amongst other things a public domain dedication, a waiver of moral rights (to the extent possible), and a fall back to a public license to try to address as many international law issues as possible.

Clearly a lot of effort has been put into getting things into the public domain over the years. (And for good reason in many cases.) But it would be interesting (and kind of cool) to see a project take itself out of the public domain by intentionally introducing code that has copyright projection available to it. And actually this is related to one of the less cited advantages (that I believe the FSF has put forward) about licensing a project under the GPL. A GPL project can take code released under the terms of many different kinds of licenses (including code not subject to copyright); but, in contrast, more permissively licensed projects are (unfortunately) intentionally closing themselves off from borrowing code from a whole portion of the software commons by refusing to accept GPL licensed code. (Sorry, I ironically don't have a citation for that sentence. My apologizes to RMS and the FSF if that is not their view.)

In that vein, instead of waiting for a copyright holder to submit a GPL licensed contribution, another approach (perhaps more aggressive) would be for the government to intentionally use a GPL licensed, or other copyleft licensed, library/code in the project. For instance, the venerable [GNU Readline library](http://cnswww.cns.cwru.edu/php/chet/readline/rltop.html) is a GPL license library. It [appears](http://clisp.cvs.sourceforge.net/viewvc/clisp/clisp/doc/Why-CLISP-is-under-GPL) that the project has for years refused requests to change the license to something more liberal because the copyright holders want other projects to adopt the GPL. If, for example, the government integrates the Readline library, or any other GPL licensed library, into the core of a project then for practical purposes the project as a whole would have to be considered a GPL work (even though, as you acknowledge, the portions contributed by federal employees would not technically have copyright protection). It would raise a interesting public policy question though about if that was consistent with the intent of Section 105.

##Contributor policies (To DCO or to CLA, that is the question? Or I dare say; might one even ask for assignment!)
@tomberek To the second point in your alternative option, I wonder if you have considered other kinds of contributor policies besides requiring the DCO. For instance your contribution policy could require people to sign a contributor agreement before you will accept their contributions to the project. The Apache Foundation's [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) (and the related Corporate CLA) is one the more popular contributor agreements in use. The [Apache license)[https://www.apache.org/licenses/] page has some more information on them. For better or worse it has been adopted by many other projects some with modifications and some without modifications.

In either case, I think it is awesome that you guys are thinking so much about how to accept contributions to government projects from outside developers. Many projects overlook this step and end up with a bit of a problem if they inadvertently start letting in contributions explicitly licensed under conflicting terms. Or even more confusingly they start using the DCO or the CLA, but don't actually right down and publish a policy that says that's what they do. And in some cases the project maintainers eventually forget and start letting in contributions without meeting the requirements of the contribution policy.

One of the things that many people like about the DCO (besides being simple enough for nonlawyers to understand) is that it allows contributors to retain their copyright and downstream users receives the license to use the that portion of contributed code in the software directly from the contributing developer. The theory being that each contributor licenses their individual portion both to the government and directly to all downstream users. You end up with multiple copyright holders in the project who have coordinated the licenses they each individual grant to allow downstream developers to use the code by complying with a single set of licensing terms. This is very much in line with how many modern FOSS licenses are explicitly designed to work (and in practice almost all FOSS licenses do work.) (Note that many modern FOSS licenses intentionally prohibit sublicensing so that end users can only take their license from the copyright holder, or licensor, directly).

In contrast the Apache CLA (and many other contributor license agreements) give the project's host organization a very broad license that includes the right for the host organization to sublicense the contribution. When those are used the contributor still retains the copyright, but it is actually the host organization that grants a license to the end user and other developers.

A third approach that is often associate with the [GNU project](https://www.fsf.org/bulletin/2014/spring/copyright-assignment-at-the-fsf) is to ask for contributors to assign copyright to the host organization. It is similar to the typical CLA approach in that it is the host organization that is granting the license to the end user. But In this case the host organization is doing so as the copyright holder, not merely as a licensee with a broad license to sublicense. The advantage of this approach is that the host organization is also the copyright holder, so in theory the host organization is in a stronger position to enforce the license by suing for copyright infringement. Considering as a general rule that the U.S. government is able to hold copyrights assigned to it, this could also be a viable approach.

I would be interested to hear if you thought the federal government's status in copyright law (or just its general way of operating bureaucratically) makes any of theses approach more desirable. All three have the potential for creating unique consequences, especially when it comes time to enforce the license if someone violates it. I don't think any of the approaches is cleared favored by the FOSS community as a whole although there are strong advocates for and against some of these.

Regardless of the approach you go with it is awesome that you guys are thinking of the long term health of your projects and the communities that might develop around them. Having a written contributor policy is huge advantage for projects and can avoid a lot of hassles. This is especially true in projects that end up accepting contributions from a large number of copyright holders. But even in relatively small projects it can help avoid disputes by making clear what the rules for working as part of the project team are.

-Marc
I am still not your lawyer, and this still wasn't legal advice. It also represents my opinions alone and does not represented the opinion of my employer.

**Comment by irl** [comment]
> I think this would be especially exciting, if a government project took the opportunity to adopt a copyleft license such as the GPL or AGPL! I have seen mentioned in other threads on in this repo that one of this group's concerns is the long-term openness of the project.

My argument against this was that this imposes restrictions on distribution which are not present under U.S. copyright law when the work was entirely the product of U.S. government employees. This is also a complex document, where the BSD/MIT licenses could almost be memorised.

**Comment by notklaatu** [comment]
> But it would be interesting (and kind of cool) to see a project take itself out of the public domain by intentionally introducing code that has copyright projection available to it. And actually this is related to one of the less cited advantages (that I believe the FSF has put forward) about licensing a project under the GPL.

Agreed. GPL is a license that guarantees that the code will remain accessible to them no matter who or what uses it (unlike a license like MIT or BSD, whereby the code could be forked, used, and then made inaccessible). Since this code is being paid for by US tax payers, it seems only fitting that they get a guarantee of persistently free code.

(Although, admittedly, either option is certainly a vast improvement over having no code, as has been the case up to...well, now.)

**Comment by irl** [comment]
@notklaatu You're missing the point entirely. The original project as produced with taxpayer money is never going to inaccessible. Beyond that point, what private people/companies do with the code is unrestricted as there can be no copyright asserted on it.

> GPL is a license that guarantees that the code will remain accessible

Yes, but it restricts the freedoms associated with the software against the spirit of the public domain nature of the original work. Those who spend their own money, not taxpayer money, to modify the code should not then be forced to make available those modifications.

I'll also point out, that if DoD continued to contribute and/or use the code, they could end up in a situation where they could not legally use the code without making their modifications freely available. The situation where changes are made that DoD does not wish to make public is not a situation that is crazy to imagine. Individual contributors would retain copyright and would potentially have title to sue. GPL may turn into a nightmare scenario here, despite the best of intentions.

**Comment by marctjones** [comment]
@irl There is no requirement under the GPL to distribute merely by using the GPL licensed software. The GPL requirements kick in when you distribute code (in source or binary form). I think most attorneys experienced in the space would agree that the government is generally free to use GPL code (including modifying it) without any requirement to share it.

**Comment by marctjones** [comment]
@irl I agree with your point about public policy issues are a important consideration. The [house report](https://www.law.cornell.edu/uscode/text/17/105?qt-us_code_temp_noupdates=1#qt-us_code_temp_noupdates) discussing section 105 of the Copyright Act does address this situation. It seems like the government (i.e. congress) thought through this issue and said "depends."

 > The bill deliberately avoids making any sort of outright, unqualified prohibition against copyright in works prepared under Government contract or grant. There may well be cases where it would be in the public interest to deny copyright in the writings generated by Government research contracts and the like; it can be assumed that, where a Government agency commissions a work for its own use merely as an alternative to having one of its own employees prepare the work, the right to secure a private copyright would be withheld. However, there are almost certainly many other cases where the denial of copyright protection would be unfair or would hamper the production and publication of important works. Where, under the particular circumstances, Congress or the agency involved finds that the need to have a work freely available outweighs the need of the private author to secure copyright, the problem can be dealt with by specific legislation, agency regulations, or contractual restrictions.

**Comment by BrandonBouier** [comment]
We've pivoted based on user feedback. Take a look at the current CONTRIBUTING.md

**Comment by fulldecent** [comment]
Since license is now outside the scope of this project, this issue appears to be out of scope. I recommend to close this issue.

**Comment by tomberek** [comment]
@marctjones Thanks for the analysis. I am especially interested in:

> I think this would be especially exciting, if a government project took the opportunity to adopt a copyleft license such as the GPL or AGPL! I have seen mentioned in other threads on in this repo that one of this group's concerns is the long-term openness of the project. The GPL and AGPL were explicitly design to guarantee that code would remain in the public domain even as it changes and receives contributions over time. (See para. 2 of the GPLv3's Preamble )

What are the sorts of projects where GPL is better from a pragmatic standpoint? Are there projects that are particularly ill-suited for GPL/AGPL due to potential licensing conflict?

**Comment by irl** [comment]
@tomberek The problem with GPL is license compatibility.

https://www.gnu.org/licenses/gpl-faq.html#WhatIsCompatible
https://www.gnu.org/licenses/gpl-faq.html#WhatDoesCompatMean
https://www.gnu.org/licenses/license-list.html#GPLIncompatibleLicenses

By adding restrictions, you can accidentally prohibit things that you didn't intend to. AGPL is the worst for this.

**Comment by tomberek** [comment]
@irl I am aware of those restrictions. My question was from the perspective of a future manager of a government-produced project. We can't expect every project manager to be an expert on licensing and understand all the nuances. So general guidelines in the FAQ would help a lot. Perhaps we say:

> use MIT/BSD/ISC for almost all cases. Only use GPL/AGPL if you know what you are doing.

I can imagine some guidelines:

> Projects which are primarily used by developers (libraries, frameworks, etc) will benefit from the minimal restrictions in ISC. Projects directly for public users will benefit from the freedom to view the source code they are interacting with.

**Comment by fulldecent** [comment]
@tomberek Considering the situation of government copyrighted code then yes you are free to choose a license for this code from the traditional off the shelf licenses. One nice website for that is https://choosealicense.com/

Learning from the private sector, MIT has been very successful and I would prefer if we recommended that first since it is the most permissive.

However if you are creating an FAQ then be sure to explain why project maintainers need a license in the first place. This is discussed in #36.

@tomberek if you need an FAQ for project maintainers please just describe the audience and what you seek to accomplish. I and the many other interested people here can help you get this started. (The current FAQ appears to be for THIS project and not the FAQ for project maintainers.)

**Comment by irl** [comment]
@tomberek While I agree in principle with the split between developer focussed (e.g. libraries) and end-user focussed (e.g. graphical applications) you also then place restrictions on what the code could be in the future (e.g. graphical applications having code refactored into a library) and thus create a ticking time bomb. You can avoid this problem with copyright assignment, but I personally refuse to contribute to projects requiring copyright assignment, and only contribute in such cases where the upstream developers will accept that I dedicate the contribution to the public domain and so they can include it in their project. That's a whole other can of worms.

If there is a team that really wants to use GPL/AGPL then they are welcome to fork the BSD/MIT code and add a GPL license on it (you can use BSD code in GPL licensed projects, but not the other way round) and then if those contributions are good enough for people to accept GPL, the project may become defacto GPL, but I think that should be a choice left up to the community as opposed to dictated by the DoD when the project was originally public domain.

**Comment by marctjones** [comment]
@tomberek  Use <my favorite license> because {it is awesome | it is simple | it was written by lawyers | it wasn't written by lawyers | it is GPL | it is not GPL | I read somewhere ...}. 

I think the best license depends on your goals. And I don't know them so I am not going to give you my advice. 

In fairness to the above arguments they are all understandable and have an appeal but I think they over simplify the decision. If you want to choose a license you I think you need to do a "multifactor analysis." (In the law school sense of the word, not the math sense. Well maybe the math sense, but I don't understand the math sense so I can't say for sure.)

Experience legal counsel can help or even one of the many excellent license consultants can tell you a lot about what you might want to consider.

Big things to consider off the top of my head are:
1) What are the licenses for the libraries are you doing to use? Are they compatible?
2) How do you feel about copyleft?
3) What is the predominant license used by the language community you are developing in? For example if you are writing a lot Java then a lot of people would say default to Apache. But if you are writing clojure code default to Eclipse (unless you are using a lot of Java libraries then go with Apache, maybe?) 
4) But that doesn't always work because if you are writing C because then maybe go with a permissive license like MIT or BSD unless you want copyleft then go with the GPL. But if you are writing C code for a Apache HTTPD module for example you probably want to go with Apache, even if you are a fan of the GPL.
5) Choose a popular license. If you are not using one of [OSI's popular licenses](https://opensource.org/licenses) then, you should probably have a good reason besides "I like it."
6) Are you particularly concerned with patents? Or Trademarks?
7) Do you want to have a separate contributor agreement or policy. Probably best to avoid a license that includes one if you want to manage it yourself.
8) "weak" or "strong" copyleft (my apologies to any LGPL fans for using those terms).
9) What is your business strategy? What do you want to happen to your code? 

This is a nonexhaustive list of possible things some people think about. I am sure there are other things you will want to consider as well. And of course, I am sure someone will tell you that my list of things to consider is wrong and they might right; except to the extent they are wrong.

Personally I am surprised more companies don't like to release their code as GPL. It makes it harder to compete with you if you release your code as GPL because your competitors can't just hide your efforts into their own proprietary project. They have to share their work or forgo the benefit of your work. For instance notice that Oracle uses copyleft licenses a lot (GPL and CDDL) but thier contributor agreement asks people to give them a very permissive license. They get all of the benefits of free labor but give less than that to their competitors.

Also look at GitHub. Sure they release software as open source under permissive licenses, but notice none of the stuff they release is a direct competitor to their main business model. It seems like they use open source as an advertising method to generate goodwill in the community, but they protect their core business model by not releasing the software most important to them  at all. In contrast, GitLab, a competitor of GitHub, releases stuff is in direct competition to their business but I believe all of the is released as copyleft. So two very different business strategies in the same small market. I would be willing to be neither companies's strategy is anything less than an intentional strategy. 

I do not think the GPL or the AGPL are scary if you think about the implications first. But you probably want experienced legal counsel to advise you on the selection of any license copyleft or otherwise. MongoDB uses the [AGPL](https://www.mongodb.com/community/licensing) apparently very successfully. Note their goal is to make it hard for someone to fork MongoDB and create an online hosted service with proprietary extensions that competes with them directly. They want you to share improvement backs to the community. I could be wrong but I think at one point in time the MongoDB folks were even willing to give you a comfort letter if you want to use the AGPL license as a backend in house. If I understand the original author's intentions correctly the AGPL was meant to stop FOSS from being co-opted by SAAS providers and force SAAS providers to share their improvements back to the community. That is exactly how MongoDB appears to be using it. So if that is a concern for you, you might want to consider it, but even then that might not be your only concern either. 

You might want to also consider the advice from the FSF. They are happy to tell you when you shouldn't use the [LGPL](https://www.gnu.org/licenses/license-recommendations.html) or the [GPL](https://www.gnu.org/licenses/license-recommendations.html). (They are also happy to tell when to use them as well ;) ) But pretty sure they have a bias as well and not everyone in the FOSS community shares their exact point of view. Is the FSF objectively wrong though? Or do people that disagree with them just have different concerns and goals?

So what license should you use? What things should you consider when selecting a license? Good questions? When is the GPL good/bad? I have no idea what the answer to any of those questions is. ;) Definitely don't listen to a random guy on the Internet though, myself included (including this sentence?) No, except for this sentence? Wait, now I don't know what sentence I am referring to anymore.

**Comment by shawoods** [comment]
@marctjones @irl Your thoughts about using copyleft licenses are helpful. We are actively triaging our projects now to find an appropriate candidate. 

>I'll also point out, that if DoD continued to contribute and/or use the code, they could end up in a situation where they could not legally use the code without making their modifications freely available. 

Agreed. That is why it's so important, as you all are saying, to carefully analyze candidates projects and determine whether there are factors that would make a strong copyleft license problematic. As an aside, we are looking at https://www.fsf.org/campaigns/priority-projects/ and discussing whether anything we're working on aligns.

**Comment by marctjones** [comment]
@shawoods Have you see the FSF FAQ page. While thier interpretation is of the GPL is not authoritative, as the authors of the GPL I think most people find thier interpretation persuasive. 

But at least according to the [FSF,](https://www.gnu.org/licenses/gpl-faq.html#GPLRequireSourcePostedPublic) you do not need to share modifications merely because you made modifications.  

> The GPL does not require you to release your modified version, or any part of it. You are free to make modifications and use them privately, without ever releasing them. This applies to organizations (including companies), too; an organization can make a modified version and use it internally without ever releasing it outside the organization.
> 
> But if you release the modified version to the public in some way, the GPL requires you to make the modified source code available to the program's users, under the GPL.
> 
> Thus, the GPL gives permission to release the modified program in certain ways, and not in other ways; but the decision of whether to release it is up to you.

**Comment by marctjones** [comment]
@shawoods Makes sense. Keep it simple.

**Comment by marctjones** [comment]
@shawoods I think this issue has evolved a few times. Maybe now is a good time to close it before it evolves again.

**Comment by shawoods** [comment]
This thread has pivoted a few times. I think all of the issues have been closed out, but if anyone disagrees we can always reopen it.

**Comment by NoahKunin** [comment]
Wow! Thanks for all the comments. Will take a hot minute to review. Thanks again.
